code/modules/mob/living/carbon/human/human_organs.dm

```diff
@@ -241,3 +241,97 @@
 	if(L)
 		L.rupture()
 
+//Registers an organ and setup the organ hierachy properly.
+//affected  : Parent organ if applicable.
+//in_place  : If true, we're performing an in-place replacement, without triggering anything related to adding the organ in-game as part of surgery or else.
+/mob/living/carbon/human/add_organ(var/obj/item/organ/O, var/obj/item/organ/external/affected = null, var/in_place = FALSE, var/update_icon = TRUE)
+	. = ..()
+	if(ispath(O.type, /obj/item/organ/internal))
+		if(!affected)
+			affected = organs_by_name[O.parent_organ]
+			if(!affected)
+				CRASH("mob/living/carbon/human/proc/add_organ(): Tried to add an internal organ to a non-existing parent external organ!")
+		internal_organs |= O
+		internal_organs_by_name[O.organ_tag] = O
+
+	else if(ispath(O.type, /obj/item/organ/external))
+		organs |= O
+		organs_by_name[O.organ_tag] = O
+		refresh_modular_limb_verbs()
```

Comment:
The reason Loaf didn't like this structure is this branching here; it seems like this should all be done in an organ proc. The stuff below the install call can presumably be moved to the `on_gained-organ` follow-up, so that doesn't leave much that would naturally belong in this main call.

Reply:
From what loaf told me, what he didn't like was the type checking. Or at least part of what he didn't like. Which makes sense. I did that to avoid having to mess with the organ lists. But yeah, its not ideal, so I've been applying some of his changes to the organ lists from his PR. Its not yet ready though.

And well, I'm not entirely sure why the chain should be started by the organs themselves? As it is currently with basically anything else, when you add something to something else, you add it via the future owner's procs. It makes sense in terms of hierarchy. 
Plus, the organs themselves are making direct changes and triggering things in the parent mob without being aware of the context. For instance, one of the common issue I've been having with the old way its done is how organs would keep triggering body or dna update at the worst possible time, just cause they feel like it. Feels like the mob initiating the chain makes it easier to deal with things like that, since the mob actually knows about its own state, while the organs, or even the caller, don't.

And there's a reason why on_install and install are separate. Another issue of the old code is that you couldn't selectively avoid/override triggering effects from organ installation/removal. Which cause problems when rejuvinating(Cause it just deleted all old organs in place and caused ui bugs and the like) or when replacing organs at runtime via code and not surgery. I try to keep the bare minimum in install, and then anything like dropping clothing, dying, and etc is done in the on_***** proc. 
Its pretty subtle, but it makes a huge difference in my opinion.

Reply:
I don't really care what starts the chain, and if you believe there's a good reason for the separation of the stuff below from `on_install` that's fine. Loaf and I don't love the istypes right above here.

I do think you should aim for 100% accuracy of the organ lists at all times, even during "temporary" installations or whatever (re: the two comments below too). That means that if the lists are modified in this proc, and you are installing/removing an organ, you need to commit to _only_ using this proc, treating the rest as internal/private.

Reply:
Don't worry, I don't like the istype either XD I've already redid it in a better way, actually using polymorphism for it. I'm currently integrating that with loaf's approach to organs lists for my next push. (Just been having issues with all human mobs working completely fine, but not spawning with eyes... I wish loaf had time to test his PR's code before XD)

And, I agree. I'm also aiming for 100% accuracy and consistency of the mob organs. 
Its just organs aren't always inside mobs. And sometimes, they can be inside mobs in a detached state (Basically the surgery code turns them into "implants" when detaching them, and essentially remove them from the organ list). So, some cases get a bit ambiguous on that front, and I want to reuse code as much as possible. I might rename the install/uninstall procs on organs to make their intended usage a bit more clear.. I'm also looking into reducing the amount of args on those procs..

I'm also considering getting rid of the auto-installation code in the organ init code, since it introduce more unknowns that I'm willing to deal with for a really dubious convenience advantage? And anyways, the only time we're creating organs actually inside a mob is in the create_missing_organ proc, and the VV menu...